### DISCUSSION TITLE
"Erroneous arithmetic operation" in EB_CNS when AMR based on density gradient is activated for 2D simulation

### DISCUSSION THREAD

**Discussion by JerryNiOnly** [AMReX-Codes/amrex#3320] [Q&A] [answered]
Hi, I wanted to share my experience using the EB_CNS examples, specifically the Sod example, and its source codes for my research. I made some adjustments to the input files and managed to reproduce the Cylinder-shock interaction test and Schardin experiment test mentioned in the article by modifying the `EB_CNS/Sod` example for the 2D scenario where `DIM=2` is set in the makefile.

The simulations worked well, except for the grids around the moving shock wave, which were not properly refined. To address this issue, I added two lines to the settings in the input file: `cns.refine_max_dengrad_lev = 3` (which was originally set to -1) and `cns.refine_dengrad = 0.005`. These settings were intended to activate adaptive mesh refinement (AMR) for grids with high density gradients. However, upon starting the simulation, it was immediately aborted, and the console displayed an "Erroneous arithmetic operation" message.

To troubleshoot this issue, I investigated the backtrace file and reviewed the relevant source code files where the error could potentially occur. I discovered that in the file `amrex/Tests/EB_CNS/Source/CNS_tagging.H`, the z-dimensional parameter `az` was not ignored even when `DIM=2` was set. The "Erroneous arithmetic operation" occurred when the max function in the following statement was executed:
```cpp
if (amrex::max(ax, ay, az) >= dengrad_threshold)
```

To resolve this problem, I implemented a fix by adding a conditional compilation if statement as shown below. In conclusion, I believe it would be beneficial to notify the AMReX team about this issue so that other users working with the 2D EB_CNS may avoid spending a significant amount of time being confused by similar errors. Thank you for your attention, and I welcome any feedback or suggestions regarding this matter.

```cpp
AMREX_GPU_DEVICE AMREX_FORCE_INLINE
void
cns_tag_denerror (int i, int j, int k,
                  amrex::Array4<char> const& tag,
                  amrex::Array4<amrex::Real const> const& rho,
                  amrex::Real dengrad_threshold, char tagval) noexcept
{
    amrex::Real ax = std::abs(rho(i+1,j,k) - rho(i,j,k));
    amrex::Real ay = std::abs(rho(i,j+1,k) - rho(i,j,k));
    ax = amrex::max(ax,std::abs(rho(i,j,k) - rho(i-1,j,k)));
    ay = amrex::max(ay,std::abs(rho(i,j,k) - rho(i,j-1,k)));
#if (AMREX_SPACEDIM == 3)
    amrex::Real az = std::abs(rho(i,j,k+1) - rho(i,j,k));
    az = amrex::max(az,std::abs(rho(i,j,k) - rho(i,j,k-1)));
    if (amrex::max(ax,ay,az) >= dengrad_threshold) {
        tag(i,j,k) = tagval;
    }
#else
     if (amrex::max(ax,ay) >= dengrad_threshold) {
        tag(i,j,k) = tagval;
    }
#endif
}
```

**Comment by WeiqunZhang** [comment, marked as answer]
Thanks for reporting this! We will fix this.
